### PR TITLE
Speed up BUFFER increases (minimize reallocs)

### PR DESCRIPTION
**Pull request by ktsaou**
Optimize BUFFER increases for large web requests.

This PR makes buffer increases about 25% faster for large buffers.

In the following test case, using the BUFFER of the master, it finishes in 65 ms. With this PR it finishes in 50 ms.

Test:

```c
int main(int argc, char **argv) {
    BUFFER *wb = buffer_create(1024);

    const char *string = "012345678901234567890123456789012345678901234567890123456789";
    size_t size_of_string = strlen(string);
    size_t total = 0;

    usec_t start = now_realtime_usec();
    while(total < 40*1024*1024) {
        buffer_sprintf(wb, "%s", string);
        total += size_of_string;
    }
    usec_t end = now_realtime_usec();
    printf("Buffer of %zu bytes finished in %llu usec\n", buffer_strlen(wb), end-start);

    log_allocations();

    buffer_free(wb);
    exit(0);
}
```

Test results from master:

```
Buffer of 41943060 bytes finished in 65844 usec
MAIN MEMORY ALLOCATIONS: (0697@daemon/main.c:main): Allocated 40961 KiB (+41944078 B), mmapped 0 KiB (+0 B): : malloc 1 (+1), calloc 1 (+1), realloc 40920 (+40920), strdup 0 (+0), free 0 (+0)
```

Test results from this PR:

```
Buffer of 41943060 bytes finished in 50036 usec
MAIN MEMORY ALLOCATIONS: (0697@daemon/main.c:main): Allocated 49152 KiB (+50331702 B), mmapped 0 KiB (+0 B): : malloc 1 (+1), calloc 1 (+1), realloc 15 (+15), strdup 0 (+0), free 0 (+0)
```

---

## real-life effect

Running a parallel test with on a context with 1000 charts, 100 dimensions each, like this:

```
siege --verbose --concurrent=255 --benchmark --reps=1 "http://localhost:19999/api/v1/data?context=example.random&before=0&after=-60&points=60&options=jsonwrap"
```

This PR provides a 10% benefit.

The master:

```
Transactions:                255 hits
Availability:             100.00 %
Elapsed time:             131.46 secs
Data transferred:        2384.86 MB
Response time:            110.83 secs
Transaction rate:           1.94 trans/sec
Throughput:            18.14 MB/sec
Concurrency:              214.98
Successful transactions:         255
Failed transactions:               0
Longest transaction:          131.46
Shortest transaction:          93.43
```

This PR:

```
Transactions:                255 hits
Availability:             100.00 %
Elapsed time:             119.09 secs
Data transferred:        2384.76 MB
Response time:             97.45 secs
Transaction rate:           2.14 trans/sec
Throughput:            20.02 MB/sec
Concurrency:              208.66
Successful transactions:         255
Failed transactions:               0
Longest transaction:          119.09
Shortest transaction:          83.44
```

### CODE REVIEW

File: libnetdata/buffer/buffer.c [hunk 1]
```diff
@@ -420,16 +427,22 @@ void buffer_increase(BUFFER *b, size_t free_size_required) {
...
-
     if(left >= free_size_required) return;
 
-    size_t increase = free_size_required - left;
-    if(increase < WEB_DATA_LENGTH_INCREASE_STEP) increase = WEB_DATA_LENGTH_INCREASE_STEP;
+    size_t wanted = free_size_required - left;
+    size_t minimum = WEB_DATA_LENGTH_INCREASE_STEP;
+    if(minimum > wanted) wanted = minimum;
+
+    size_t optimal = b->size * 2;
```
Comment (vkalintiris) [inline libnetdata/buffer/buffer.c:436]: We will end up with 3x the memory that we had before the increase, right? This is a little bit too much. The amortized cost will be the same if we grow by a factor of 0.5x to 1x.

Comment (ktsaou) [reply libnetdata/buffer/buffer.c:436]: this happens only for buffers < 1M

Comment (ktsaou) [reply libnetdata/buffer/buffer.c]: I accepted your suggestion to go double the buffer size.

**Review by thiagoftsm — APPROVED** [review]
I tested this PR and I compare with master branch, and I can confirm improvement in our stats. 
I also observe that cloud was working fine and the host was visible all the time.
LGTM!

**Review by MrZammler — APPROVED** [review]

**Pull request merged**